go-controller/pkg/ovn/base_network_controller_secondary.go

```diff
@@ -0,0 +1,268 @@
+		// the pod is not attached to this specific network
+		klog.V(5).Infof("Pod %s/%s is not attached on this network controller %s error (%v) ",
+			pod.Namespace, pod.Name, bsnc.GetNetworkName(), err)
+		return nil
+	}
+
+	if bsnc.lsManager.IsNonHostSubnetSwitch(switchName) {
+		return nil
+	}
+
+	nadName := util.GetNADName(network.Namespace, network.Name)
+
+	var libovsdbExecuteTime time.Duration
+
+	start := time.Now()
+	defer func() {
+		klog.Infof("[%s/%s] addLogicalPort for NAD %s took %v, libovsdb time %v",
+			pod.Namespace, pod.Name, nadName, time.Since(start), libovsdbExecuteTime)
+	}()
+
+	ops, lsp, podAnnotation, newlyCreated, err := bsnc.addLogicalPortToNetwork(pod, nadName, network)
+	if err != nil {
+		return err
+	}
+
+	recordOps, txOkCallBack, _, err := bsnc.AddConfigDurationRecord("pod", pod.Namespace, pod.Name)
+	if err != nil {
+		klog.Errorf("Config duration recorder: %v", err)
+	}
+	ops = append(ops, recordOps...)
+
+	transactStart := time.Now()
+	_, err = libovsdbops.TransactAndCheckAndSetUUIDs(bsnc.nbClient, lsp, ops)
+	libovsdbExecuteTime = time.Since(transactStart)
+	if err != nil {
+		return fmt.Errorf("error transacting operations %+v: %v", ops, err)
+	}
+	txOkCallBack()
+	bsnc.podRecorder.AddLSP(pod.UID, bsnc.NetInfo)
+
+	// if somehow lspUUID is empty, there is a bug here with interpreting OVSDB results
+	if len(lsp.UUID) == 0 {
+		return fmt.Errorf("UUID is empty from LSP: %+v", *lsp)
+	}
+
+	_ = bsnc.logicalPortCache.add(pod, switchName, nadName, lsp.UUID, podAnnotation.MAC, podAnnotation.IPs)
+
+	if newlyCreated {
+		metrics.RecordPodCreated(pod, bsnc.NetInfo)
+	}
+	return nil
+}
+
+// removePodForSecondaryNetwork tried to tear down a for on a secondary network. It returns nil on success
+// and error on failure; failure indicates the pod tear down should be retried later.
+func (bsnc *BaseSecondaryNetworkController) removePodForSecondaryNetwork(pod *kapi.Pod, portInfoMap map[string]*lpInfo) error {
+	var nadName string
+	var portInfo *lpInfo
+
+	podDesc := pod.Namespace + "/" + pod.Name
+	klog.Infof("Deleting pod: %s for network %s", podDesc, bsnc.GetNetworkName())
+
+	if portInfoMap != nil {
+		if len(portInfoMap) == 0 {
+			return nil
+		}
+		for nadName, portInfo = range portInfoMap {
+			break
+		}
+	} else {
+		if !util.PodWantsNetwork(pod) {
+			return nil
+		}
+		if !util.PodScheduled(pod) {
+			return nil
+		}
+
+		// podInfoMap does not exist, it means the pod failed to be created or this is called when ovnkube-master
+		// restarts and logicalPortCache is not filled in. In this case, we'd still need to delete resources (IPs,
+		// logical switch port) created for this Pod.
+
+		// get all nadNames from pod annotation, and see if any NAD is of this network
+		podNetworks, err := util.UnmarshalPodAnnotationAllNetworks(pod.Annotations)
+		if err != nil {
+			return err
+		}
+
+		for nadName = range podNetworks {
+			if bsnc.HasNAD(nadName) {
+				break
+			}
+		}
+		if nadName == "" {
+			return nil
+		}
```

Comment:
What if the NAD has been removed from the network controller before we get here? Can that happen?
For example, both the NAD and the Pod were deleted before the restart or after the add failed.
Wouldn't it be better that any controller that found the annotation would just blindly attempt to delete it's resources instead of risking that no controller does?
Would it help if the key to the annotation where `network/nad` instead of only `nad`?

Reply:
for both cases, the logical switch and all logical switch port on those switches will be deleted as the results of NAD deletion. We discussed this in the meeting and agreed that we could probably prevent it from happening or/and add NADUUID or netconf external_ids to minimize the impact. I personally do not like the idea that doing something for other networks. What do you think?